internal/stacks/stackruntime/internal/stackeval/local_value_config.go

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package stackeval
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform/internal/promising"
+	"github.com/hashicorp/terraform/internal/stacks/stackaddrs"
+	"github.com/hashicorp/terraform/internal/stacks/stackconfig"
+	"github.com/hashicorp/terraform/internal/stacks/stackplan"
+	"github.com/hashicorp/terraform/internal/tfdiags"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// LocalValueConfig represents a "variable" block in a stack configuration.
+type LocalValueConfig struct {
+	addr   stackaddrs.ConfigLocalValue
+	config *stackconfig.LocalValue
+
+	main *Main
+
+	validatedValue perEvalPhase[promising.Once[withDiagnostics[cty.Value]]]
+}
+
+var (
+	_ Validatable          = (*LocalValueConfig)(nil)
+	_ Referenceable        = (*LocalValueConfig)(nil)
+	_ namedPromiseReporter = (*LocalValueConfig)(nil)
+)
+
+func newLocalValueConfig(main *Main, addr stackaddrs.ConfigLocalValue, config *stackconfig.LocalValue) *LocalValueConfig {
+	if config == nil {
+		panic("newLocalValueConfig with nil configuration")
+	}
+	return &LocalValueConfig{
+		addr:   addr,
+		config: config,
+		main:   main,
+	}
+}
+
+func (v *LocalValueConfig) Addr() stackaddrs.ConfigLocalValue {
+	return v.addr
+}
+
+func (v *LocalValueConfig) tracingName() string {
+	return v.Addr().String()
+}
+
+func (v *LocalValueConfig) Declaration() *stackconfig.LocalValue {
+	return v.config
+}
+
+// StackConfig returns the stack configuration that this input variable belongs
+// to.
+func (v *LocalValueConfig) StackConfig(ctx context.Context) *StackConfig {
+	return v.main.mustStackConfig(ctx, v.Addr().Stack)
+}
+
+// ExprReferenceValue implements Referenceable
+func (v *LocalValueConfig) ExprReferenceValue(ctx context.Context, phase EvalPhase) cty.Value {
+	return cty.StringVal("parent")
```

Comment:
What does this string represent? 🤔

Reply:
This was for testing if we encounter this path somewhere, but it didn't seem like it showed up, so I've replaced it with DynamicVal seeing as how the type was stored for referencing later. I wonder if that's correct as well. Would love to learn more about whether this interface implementation is required for locals and how it's used in the scheme of things.

Reply:
I believe that `ExprReferenceValue` is called on descendants of `StackConfig` during validation. Some configuration constructs are able to give more informative results than just an unknown value of unknown type (e.g. embedded stack calls). It seems valid to me to return that in this case since we won't always know anything about the value or its type here.

Reply:
If I'm understanding the rest of `local_value_config.go` correctly, I think a plausible implementation of `ExprReferenceValue` here would be to call `v.Validate` value and then return its first return value, discarding the diagnostics.

That would then return whatever is our best approximation of the local value based on the information we have during validation. In the best case where the local value is defined as a literal value it would be that value. If it were defined as a reference to something else we don't know during validation then it would instead be an unknown value placeholder for that thing.

Reply:
(`validateValueInner` should probably also take a `phase EvalPhase` argument and pass it in to the expression evaluator, rather than hard-coding `ValidatePhase`, because these config eval functions also get called during `PlanPhase` as a static check before we do the dynamic evaluation. It might make things clearer to rename `ValidateValue` and `validateValueInner` to be `CheckValue` and `checkValueInner` to avoid the implication that they are used only during the validate phase; I suspect that naming was a remnant of some early work when it wasn't clear to me yet exactly how we'd handle plan-time static checks.)